modules/channel-vonage/src/backend/client.ts

```diff
@@ -0,0 +1,362 @@
+  private logger: sdk.Logger
+  private vonage: Vonage
+  private webhookUrl: string
+  private conversations: sdk.experimental.conversations.BotConversations
+  private messages: sdk.experimental.messages.BotMessages
+  private kvs: sdk.KvsService
+
+  constructor(
+    private bp: typeof sdk,
+    private botId: string,
+    private config: Config,
+    private router: sdk.http.RouterExtension,
+    private baseRoute: string
+  ) {
+    this.logger = bp.logger.forBot(botId)
+    this.conversations = this.bp.experimental.conversations.forBot(this.botId)
+    this.messages = this.bp.experimental.messages.forBot(this.botId)
+    this.kvs = this.bp.kvs.forBot(this.botId)
+  }
+
+  async initialize() {
+    if (!this.config.apiKey || !this.config.apiSecret || !this.config.applicationId || !this.config.privateKey) {
+      return this.logger.error(
+        `[${this.botId}] The apiKey, apiSecret, applicationId and privateKey must be configured in order to use this channel.`
+      )
+    }
+
+    const url = (await this.router.getPublicPath()) + this.baseRoute
+    this.webhookUrl = url.replace('BOT_ID', this.botId)
+
+    // Doc: https://developer.nexmo.com/messages/concepts/messages-api-sandbox
+    this.vonage = new Vonage(
+      {
+        apiKey: this.config.apiKey,
+        apiSecret: this.config.apiSecret,
+        applicationId: this.config.applicationId,
+        privateKey: Buffer.from(this.config.privateKey) as any,
+        signatureSecret: this.config.signatureSecret
+      },
+      {
+        debug: this.config.useTestingApi ? true : false,
+        apiHost: this.config.useTestingApi ? ApiBaseUrl.TEST : ApiBaseUrl.PROD
+      }
+    )
+
+    if (this.config.useTestingApi) {
+      this.logger.info('Vonage configured to use the testing API!')
+    }
+
+    this.logger.info(`Vonage inbound webhooks listening at ${this.webhookUrl}/inbound`)
+    this.logger.info(`Vonage status webhooks listening at ${this.webhookUrl}/status`)
+  }
+
+  async handleWebhookRequest(body: VonageRequestBody) {
+    debugIncoming('Received message', body)
+
+    const userId = body.from.number
+    const botPhoneNumber = body.to.number
+
+    const conversation = await this.conversations.recent(userId)
+    await this.kvs.set(`vonage-number-${conversation.id}`, { botPhoneNumber })
+
+    // https://developer.nexmo.com/api/messages-olympus?theme=dark
+    let payload: sdk.IO.IncomingEvent['payload'] = null
+    switch (body.message.content.type) {
+      case 'text':
+        const text = body.message.content.text
+        payload = { type: 'text', text }
+
+        const index = Number(text)
+        if (index) {
+          payload = (await this.handleIndexResponse(index - 1, userId, conversation.id)) ?? payload
+
+          if (payload.type === 'url') {
+            return
+          }
+        }
+
+        await this.kvs.delete(this.getKvsKey(userId, conversation.id))
+        break
+      case 'audio':
+        const audio = body.message.content.audio.url
```

Comment:
Exactly! If you send an audio file through WhatsApp, the value of `body.message.content.type` will be 'file' not 'audio'

Reply:
Never mind what I just said. You can still drag and drop the audio file and it will be sent as 'audio'. When you send a document, the type is 'file'. I thought you could only send audio files as documents since this is the only option available in the attachment menu (I'm using the web version though!)

Vonage's API does not differentiate the two even though WhatsApp seems to!

![Screenshot from 2021-04-21 18-05-30](https://user-images.githubusercontent.com/9640576/115626971-3fa8fe80-a2cc-11eb-9c30-29c09bb0392d.png)